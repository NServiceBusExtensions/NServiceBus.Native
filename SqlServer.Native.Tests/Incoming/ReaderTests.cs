using System.Collections.Generic;
using NServiceBus.Transport.SqlServerNative;
using ObjectApproval;
using Xunit;
using Xunit.Abstractions;

public class ReaderTests : TestBase
{
    string table = "ReaderTests";

    [Fact]
    public void Single_bytes()
    {
        TestDataBuilder.SendData(table);
        var reader = new QueueManager(table, SqlConnection);
        var result = reader.ReadBytes(1).Result;
        ObjectApprover.VerifyWithJson(result);
    }

    [Fact]
    public void Single_bytes_nulls()
    {
        TestDataBuilder.SendNullData(table);
        var reader = new QueueManager(table, SqlConnection);
        var result = reader.ReadBytes(1).Result;
        ObjectApprover.VerifyWithJson(result);
    }

    [Fact]
    public void Single_stream()
    {
        TestDataBuilder.SendData(table);
        var reader = new QueueManager(table, SqlConnection);
        using (var result = reader.ReadStream(1).Result)
        {
            ObjectApprover.VerifyWithJson(result.ToVerifyTarget());
        }
    }

    [Fact]
    public void Single_stream_nulls()
    {
        TestDataBuilder.SendNullData(table);
        var reader = new QueueManager(table, SqlConnection);
        using (var result = reader.ReadStream(1).Result)
        {
            ObjectApprover.VerifyWithJson(result.ToVerifyTarget());
        }
    }

    [Fact]
    public void Batch_bytes()
    {
        TestDataBuilder.SendMultipleData(table);

        var reader = new QueueManager(table, SqlConnection);
        var messages = new List<IncomingBytesMessage>();
        var result = reader.ReadBytes(size: 3,
                startRowVersion: 2,
                action: message => { messages.Add(message); })
            .Result;
        Assert.Equal(4, result.LastRowVersion);
        Assert.Equal(3, result.Count);
        ObjectApprover.VerifyWithJson(messages);
    }

    [Fact]
    public void Batch_stream()
    {
        TestDataBuilder.SendMultipleData(table);

        var reader = new QueueManager(table, SqlConnection);
        var messages = new List<object>();
        var result = reader.ReadStream(size: 3,
                startRowVersion: 2,
                action: message => { messages.Add(message.ToVerifyTarget()); })
            .Result;
        Assert.Equal(4, result.LastRowVersion);
        Assert.Equal(3, result.Count);
        ObjectApprover.VerifyWithJson(messages);
    }

    public ReaderTests(ITestOutputHelper output) : base(output)
    {
        SqlConnection.DropTable(null, table).Await();
        var manager = new QueueManager(table, SqlConnection);
        manager.Create().Await();
    }
}