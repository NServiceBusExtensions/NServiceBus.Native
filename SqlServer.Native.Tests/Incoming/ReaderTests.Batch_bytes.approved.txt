[
  {
    "Id": "00000000-0000-0000-0000-000000000002",
    "RowVersion": 2,
    "Headers": "{}"
  },
  {
    "Id": "00000000-0000-0000-0000-000000000003",
    "RowVersion": 3,
    "CorrelationId": "theCorrelationId",
    "ReplyToAddress": "theReplyToAddress",
    "Expires": "2000-01-01T01:01:01",
    "Headers": "headers",
    "Body": "e30="
  },
  {
    "Id": "00000000-0000-0000-0000-000000000004",
    "RowVersion": 4,
    "Headers": "{}"
  }
]